applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_KS_max_stress_response_function.cpp

```diff
@@ -0,0 +1,251 @@
+// KRATOS  ___|  |                   |                   |
+//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
+//             | |   |    |   | (    |   |   | |   (   | |
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:		 BSD License
+//					 license: structural_mechanics_application/license.txt
+//
+//  Main authors:    E. Wehrle
+//                   based on max stress from Baumgaertner Daniel, https://github.com/dbaumgaertner
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "adjoint_KS_max_stress_response_function.h"
+#include "utilities/compare_elements_and_conditions_utility.h"
+
+namespace Kratos
+{
+    AdjointKSMaxStressResponseFunction::AdjointKSMaxStressResponseFunction(ModelPart& rAdjointModelPart, Parameters ResponseSettings)
+    : AdjointStructuralResponseFunction(rAdjointModelPart, ResponseSettings),
+      mrAdjointModelPart(rAdjointModelPart),
+      mCriticalPartName(ResponseSettings["critical_part_name"].GetString())
+    {
+        mTracedStressType = StressResponseDefinitions::ConvertStringToTracedStressType(ResponseSettings["stress_type"].GetString());
+        mStressTreatment = StressResponseDefinitions::ConvertStringToStressTreatment(ResponseSettings["stress_treatment"].GetString());
+
+        if(ResponseSettings.Has("echo_level"))
+            mEchoLevel = ResponseSettings["echo_level"].GetInt();
+
+        KRATOS_ERROR_IF(mStressTreatment != StressTreatment::Mean)
+            << "AdjointKSMaxStressResponseFunction::AdjointKSMaxStressResponseFunction: Specified stress treatment not supported: " << ResponseSettings["stress_type"].GetString() << std::endl;
+    }
+
+    // primal analysis for max stress approximated via modified Kreisselmeier-Steinhauser aggregation
+    AdjointKSMaxStressResponseFunction::~AdjointKSMaxStressResponseFunction(){}
+
+    double AdjointKSMaxStressResponseFunction::CalculateValue(ModelPart& rPrimalModelPart)
+    {
+        KRATOS_TRY;
+
+        ModelPart& primal_agglomeration_part = rPrimalModelPart.GetSubModelPart(mCriticalPartName);
+
+        IndexType elem_id_at_max = 0;
+
+        for(auto& elem : primal_agglomeration_part.Elements())
```

Comment:
I would do a paralel loop

Reply:
I would postpone this and do this all at once with the other responses here.